wgpu-hal/src/dx12/shader_compilation.rs

```diff
@@ -0,0 +1,260 @@
+            )
+        }
+    }
+}
+
+// The Dxc implementation is behind a feature flag so that users who don't want to use dxc can disable the feature.
+#[cfg(feature = "dxc_shader_compiler")]
+mod shader {
+    use std::path::PathBuf;
+
+    pub(crate) struct DxcContainer {
+        pub compiler: hassle_rs::DxcCompiler,
+        pub library: hassle_rs::DxcLibrary,
+        // Has to be held onto for the lifetime of the device otherwise shaders will fail to compile
+        _dxc: hassle_rs::Dxc,
+    }
+
+    pub(crate) fn get_dxc_container(
+        dxc_path: Option<PathBuf>,
+        dxil_path: Option<PathBuf>,
+    ) -> Result<Option<DxcContainer>, crate::DeviceError> {
+        // Make sure that dxil.dll exists.
+        match hassle_rs::Dxil::new(dxil_path) {
+            Ok(_) => (),
+            Err(e) => {
+                log::warn!("Failed to load dxil.dll. Defaulting to Fxc instead: {}", e);
+                return Ok(None);
+            }
+        };
+
+        let dxc = match hassle_rs::Dxc::new(dxc_path) {
+            Ok(dxc) => dxc,
+            Err(e) => {
+                log::warn!(
+                    "Failed to load dxcompiler.dll. Defaulting to Fxc instead: {}",
+                    e
+                );
+                return Ok(None);
+            }
+        };
+        let dxc_compiler = dxc.create_compiler()?;
+        let dxc_library = dxc.create_library()?;
+
+        Ok(Some(DxcContainer {
+            _dxc: dxc,
+            compiler: dxc_compiler,
+            library: dxc_library,
+        }))
+    }
+
+    pub(crate) fn compile_dxc(
+        device: &crate::dx12::Device,
+        source: &str,
+        source_name: &str,
+        raw_ep: &str,
+        stage_bit: wgt::ShaderStages,
+        full_stage: String,
+        dxc_container: &DxcContainer,
+    ) -> (
+        Result<crate::dx12::CompiledShader, crate::PipelineError>,
+        log::Level,
+    ) {
+        let mut compile_flags = arrayvec::ArrayVec::<&str, 3>::new_const();
+        compile_flags.push("-Ges"); // d3dcompiler::D3DCOMPILE_ENABLE_STRICTNESS
+        if device
+            .private_caps
+            .instance_flags
+            .contains(crate::InstanceFlags::DEBUG)
+        {
+            compile_flags.push("-Zi"); /* d3dcompiler::D3DCOMPILE_SKIP_OPTIMIZATION */
+            compile_flags.push("-Od"); /* d3dcompiler::D3DCOMPILE_DEBUG */
+        }
+
+        let blob = match dxc_container
+            .library
+            .create_blob_with_encoding_from_str(source)
+            .map_err(|e| crate::PipelineError::Linkage(stage_bit, format!("DXC blob error: {}", e)))
+        {
+            Ok(blob) => blob,
+            Err(e) => return (Err(e), log::Level::Error),
+        };
+
+        let compiled = dxc_container.compiler.compile(
+            &blob,
+            source_name,
+            raw_ep,
+            &full_stage,
+            &compile_flags,
+            None, // TODO: idk what the include handler does
```

Comment:
I still don't know what the include handler does. @MarijnS95, could you clarify what the include handler does when compiling with dxc?

Reply:
The include handler allows you to specify `#include`s in your hlsl and the handler will provide the actual text of the include.